### PR TITLE
Standardize filename regexes

### PR DESCRIPTION
**Pull request by maxbeatty**
Currently, the regex used by `ember` and `typescript` only allows filenames like `foo-bar.d.js`. The `react` regex allows for multiple dots and expanded suffixes such as `apple.stories.test.jsx`. This PR abstracts the regex being used by the three configurations making it testable.

As a bonus, I tested the `react` component pascal regex.

### CODE REVIEW

File: tests/lib/utils/filenames.js [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+
+describe('filenames', () => {
+  // filenames/match-regex ignores file extension so these tests also do
+  // https://github.com/selaux/eslint-plugin-filenames/blob/1.3.0/lib/rules/match-regex.js#L29
+  // https://github.com/selaux/eslint-plugin-filenames/blob/1.3.0/lib/common/parseFilename.js#L10
+  describe('regex', () => {
+    it('supports kebab case', () => {
+      const re = new RegExp(filenames.regex.kebab);
+
+      [
```
Comment (bmish) [inline tests/lib/utils/filenames.js:14]: Can you add a test to ensure hidden files (starting with `.`) are allowed? For files such as: `.eslintrc.js`, `.template-lintrc.js`, etc. I had allowed that in the old regexp for kebab-case.

File: lib/config/typescript.js [hunk 2]
```diff
@@ -10,15 +10,13 @@ module.exports = {
...
   ],
-  plugins: ['filenames', 'square'],
+  plugins: ['square'],
   rules: {
     // Optional eslint rules:
     'no-console': 'error',
     'sort-keys': ['error', 'asc', { natural: true }],
     'sort-vars': 'error',
 
-    'filenames/match-regex': ['error', '^.?[a-z0-9-]+(.d)?$'], // Kebab-case.
```
Comment (bmish) [inline lib/config/typescript.js:20]: This is a breaking change. Can we split this part into a separate PR (so we can consider it separately / release it separately / revert it if needed)?

File: lib/config/base.js [hunk 3]
```diff
@@ -83,6 +85,8 @@ module.exports = {
     // eslint-comments:
     'eslint-comments/no-unused-disable': 'error',
 
+    'filenames/match-regex': ['error', filenames.regex.kebab],
```
Comment (bmish) [inline lib/config/base.js:88]: This is a breaking change. Can we split this part into a separate PR (so we can consider it separately / release it separately / revert it if needed)?

Comment (maxbeatty) [reply lib/config/base.js:88]: Makes sense. I'll change back to each sub-config using the rule. Then when that is merged, I'll create a follow up PR to move it to `base` where it can gather more feedback.

**Comment by bmish** [comment]
This is awesome!
